myhpom/tasks.py

```diff
@@ -66,17 +69,103 @@ def run(self, document_url_id, document_host=None):
                 % (json.dumps(cf_run.post_data, indent=2), cf_run.response_content)
             )
 
-        # following throws an error (as it should) if response.content is not json-parsable
-        try:
-            cf_run.save_response_content(response.content)
-        except ValueError:
-            cf_run.status = CloudFactoryDocumentRun.STATUS_UNPROCESSABLE
-            cf_run.save()
-            raise
+        # following throws an error (as it should) if response.content is not json-parsable.
+        cf_run.save_response_content(response.content)
 
         return cf_run.pk
 
 
+@shared_task
+class CloudFactoryUpdateDocumentRun(Task):
+    def run(self, cf_run_id):
+        """Update the CloudFactoryDocumentRun object from the CF API.
+
+        ## Use Cases
+        * The admin user can update the status of all or selected runs from CloudFactory.
+
+        ## Parameters
+        * cf_run_id = the id (not the run_id) of the CloudFactoryDocumentRun object.
+        """
+        cf_run = CloudFactoryDocumentRun.objects.get(id=cf_run_id)
+
+        if cf_run.status in [
+            CloudFactoryDocumentRun.STATUS_PROCESSING,  # still going last we checked
+            CloudFactoryDocumentRun.STATUS_TIMEOUT,  # didn't work last time; was it created?
+            CloudFactoryDocumentRun.STATUS_ABORTED,  # we previously aborted; did it take?
+        ]:
+            try:
+                run_url = "%s/runs/%s" % (settings.CLOUDFACTORY_API_URL, cf_run.run_id)
+                response = requests.get(run_url)
+            except requests.exceptions.ConnectTimeout:
+                cf_run.status = CloudFactoryDocumentRun.STATUS_TIMEOUT
+
+            if response.status_code == 404:
+                cf_run.status = CloudFactoryDocumentRun.STATUS_NOTFOUND
+                cf_run.response_content = response.content  # not json, shouldn't raise Exception.
+                cf_run.save()
+            elif response.status_code == 200:
+                cf_run.save_response_content(response.content)
+            else:
+                # don't change the run status here -- we haven't learned anything about it,
+                # and maybe there's a response status that we don't know about
+                raise ValueError(
```

Comment:
We don't know what it is, but its definitely not expected. I'd suggest putting this in some not-okay terminal state.

Reply:
Okay, we could use STATUS_ERROR for this.